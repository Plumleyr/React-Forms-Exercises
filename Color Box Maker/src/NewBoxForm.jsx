import {useState} from 'react'

const NewBoxFrom = ({addBox}) => {
  const INITIAL_STATE = {width: '', height: '', backgroundColor: ''}

  const [formData, setFormData] = useState(INITIAL_STATE)

  const handleChange = (e) => {
    const{ name, value } = e.target;
    setFormData( formData => ({
      ...formData,
      [name]: value
    }))
  }

  const handleSubmit = (e) => {
    e.preventDefault();
    addBox(formData)
    setFormData(INITIAL_STATE)
  }

  return(
    <form onSubmit={handleSubmit}>
      <label htmlFor="width">Width:</label>
      <input 
        id='width'
        type="text" 
        name='width' 
        placeholder='width'
        value={formData.width} 
        onChange={handleChange}
      />

      <label htmlFor="height">height:</label>
      <input 
        id='height'
        type="text" 
        name='height' 
        placeholder='height'
        value={formData.height} 
        onChange={handleChange}
      />

      <label htmlFor="backgroundColor">backgroundColor:</label>
      <input 
        id='backgroundColor'
        type="text" 
        name='backgroundColor' 
        placeholder='backgroundColor'
        value={formData.backgroundColor} 
        onChange={handleChange}
      />
      <button>Create Box</button>
    </form>
  )
}

export default NewBoxFrom;