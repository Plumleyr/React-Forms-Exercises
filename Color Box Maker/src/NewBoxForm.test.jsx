import { render, screen, fireEvent } from "@testing-library/react";
import NewBoxForm from "./NewBoxForm";

test("it renders without breaking", () => {
  render(<NewBoxForm />);
});

test("it should match snapshot", () => {
  const { asFragment } = render(<NewBoxForm />);
  expect(asFragment()).toMatchSnapshot();
});

test("calls addBox with the correct data when form is submitted", () => {
  const addBoxMock = vi.fn();

  render(<NewBoxForm addBox={addBoxMock} />);

  fireEvent.change(screen.getByPlaceholderText("width"), {
    target: { value: "100px" },
  });
  fireEvent.change(screen.getByPlaceholderText("height"), {
    target: { value: "100px" },
  });
  fireEvent.change(screen.getByPlaceholderText("backgroundColor"), {
    target: { value: "blue" },
  });

  fireEvent.click(screen.getByText("Create Box"));

  expect(addBoxMock).toHaveBeenCalledTimes(1);
  expect(addBoxMock).toHaveBeenCalledWith({
    width: "100px",
    height: "100px",
    backgroundColor: "blue",
  });
});
