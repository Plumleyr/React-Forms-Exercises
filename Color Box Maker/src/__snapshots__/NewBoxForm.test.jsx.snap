// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`it should match snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="width"
    >
      Width:
    </label>
    <input
      id="width"
      name="width"
      placeholder="width"
      type="text"
      value=""
    />
    <label
      for="height"
    >
      height:
    </label>
    <input
      id="height"
      name="height"
      placeholder="height"
      type="text"
      value=""
    />
    <label
      for="backgroundColor"
    >
      backgroundColor:
    </label>
    <input
      id="backgroundColor"
      name="backgroundColor"
      placeholder="backgroundColor"
      type="text"
      value=""
    />
    <button>
      Create Box
    </button>
  </form>
</DocumentFragment>
`;
